includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php

```diff
@@ -1666,8 +1666,8 @@ private function check_attr_spec_rule_valid_url( DOMElement $node, $attr_name, $
 	 * @return string|null Protocol without colon if matched. Otherwise null.
 	 */
 	private function parse_protocol( $url ) {
-		if ( preg_match( '#^[^/]+(?=:)#', $url, $matches ) ) {
-			return $matches[0];
+		if ( preg_match( '#^\s*([^/]+)(?=:)#', $url, $matches ) ) {
```

Comment:
Is this not redundant? If the `$url` has been trimmed already, then there would never be leading whitespace, right?

Reply:
Hm, I think that change is useful now, but we could make it redundant with:

```diff
diff --git a/includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php b/includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php
index d9dd93ac8..764356573 100644
--- a/includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php
+++ b/includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php
@@ -1689,7 +1689,7 @@ class AMP_Tag_And_Attribute_Sanitizer extends AMP_Base_Sanitizer {
                if ( isset( $attr_spec_rule[ AMP_Rule_Spec::VALUE_URL ][ AMP_Rule_Spec::ALLOWED_PROTOCOL ] ) ) {
                        if ( $node->hasAttribute( $attr_name ) ) {
                                foreach ( $this->extract_attribute_urls( $node->getAttributeNode( $attr_name ) ) as $url ) {
-                                       $url_scheme = $this->parse_protocol( $url );
+                                       $url_scheme = $this->parse_protocol( trim( $url ) );
                                        if ( isset( $url_scheme ) && ! in_array( strtolower( $url_scheme ), $attr_spec_rule[ AMP_Rule_Spec::VALUE_URL ][ AMP_Rule_Spec::ALLOWED_PROTOCOL ], true ) ) {
                                                return AMP_Rule_Spec::FAIL;
                                        }
@@ -1701,7 +1701,7 @@ class AMP_Tag_And_Attribute_Sanitizer extends AMP_Base_Sanitizer {
                                foreach ( $attr_spec_rule[ AMP_Rule_Spec::ALTERNATIVE_NAMES ] as $alternative_name ) {
                                        if ( $node->hasAttribute( $alternative_name ) ) {
                                                foreach ( $this->extract_attribute_urls( $node->getAttributeNode( $alternative_name ), $attr_name ) as $url ) {
-                                                       $url_scheme = $this->parse_protocol( $url );
+                                                       $url_scheme = $this->parse_protocol( trim( $url ) );
                                                        if ( isset( $url_scheme ) && ! in_array( strtolower( $url_scheme ), $attr_spec_rule[ AMP_Rule_Spec::VALUE_URL ][ AMP_Rule_Spec::ALLOWED_PROTOCOL ], true ) ) {
                                                        
        return AMP_Rule_Spec::FAIL;
```

or:
```diff
diff --git a/includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php b/includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php
index d9dd93ac8..6eea4d0a8 100644
--- a/includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php
+++ b/includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php
@@ -1666,7 +1666,7 @@ class AMP_Tag_And_Attribute_Sanitizer extends AMP_Base_Sanitizer {
         * @return string|null Protocol without colon if matched. Otherwise null.
         */
        private function parse_protocol( $url ) {
-               if ( preg_match( '#^\s*([^/]+):#', $url, $matches ) ) {
+               if ( preg_match( '#^[^/]+(?=:)#', trim( $url ), $matches ) ) {
                        return $matches[1];
                }
                return null;
```

Reply:
I think only the former is needed. The change to allow whitespace in the regex is not needed if the `$url` is trimmed up front.

Reply:
Sounds good, I'll change this.

Reply:
It's changed in https://github.com/ampproject/amp-wp/pull/4166/commits/761981e1836126d2591675acea4c0b1b296e18c7